pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_controller.go

```diff
@@ -0,0 +1,198 @@
+package controllers
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	kube_apps "k8s.io/api/apps/v1"
+	kube_core "k8s.io/api/core/v1"
+	kube_apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kube_runtime "k8s.io/apimachinery/pkg/runtime"
+	kube_types "k8s.io/apimachinery/pkg/types"
+	kube_ctrl "sigs.k8s.io/controller-runtime"
+	kube_client "sigs.k8s.io/controller-runtime/pkg/client"
+	gatewayapi "sigs.k8s.io/gateway-api/apis/v1alpha2"
+
+	mesh_proto "github.com/kumahq/kuma/api/mesh/v1alpha1"
+	"github.com/kumahq/kuma/pkg/core"
+	core_mesh "github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
+	"github.com/kumahq/kuma/pkg/core/resources/manager"
+	"github.com/kumahq/kuma/pkg/core/resources/model"
+	"github.com/kumahq/kuma/pkg/core/resources/store"
+	k8s_common "github.com/kumahq/kuma/pkg/plugins/common/k8s"
+	"github.com/kumahq/kuma/pkg/plugins/resources/k8s"
+	mesh_k8s "github.com/kumahq/kuma/pkg/plugins/resources/k8s/native/api/v1alpha1"
+	"github.com/kumahq/kuma/pkg/plugins/runtime/k8s/metadata"
+	k8s_util "github.com/kumahq/kuma/pkg/plugins/runtime/k8s/util"
+	"github.com/kumahq/kuma/pkg/plugins/runtime/k8s/webhooks/injector"
+	util_k8s "github.com/kumahq/kuma/pkg/util/k8s"
+)
+
+const controllerName = "gateways.kuma.io/controller"
+
+// GatewayReconciler reconciles a GatewayAPI Gateway object.
+type GatewayReconciler struct {
+	kube_client.Client
+	Reader kube_client.Reader
+	Log    logr.Logger
+
+	Scheme          *kube_runtime.Scheme
+	Converter       k8s_common.Converter
+	SystemNamespace string
+	Injector        injector.KumaInjector
+	ResourceManager manager.ResourceManager
+}
+
+// Reconcile handles transforming a gateway-api Gateway into a Kuma Gateway and
+// managing the status of the gateway-api objects.
+func (r *GatewayReconciler) Reconcile(ctx context.Context, req kube_ctrl.Request) (kube_ctrl.Result, error) {
+	gateway := &gatewayapi.Gateway{}
+	if err := r.Get(ctx, req.NamespacedName, gateway); err != nil {
+		if kube_apierrs.IsNotFound(err) {
+			err := r.ResourceManager.Delete(ctx, core_mesh.NewGatewayResource(), store.DeleteByKey(req.Name, req.Name))
+			if err == nil || store.IsResourceNotFound(err) {
+				return kube_ctrl.Result{}, nil
+			}
+
+			return kube_ctrl.Result{}, err
+		}
+
+		return kube_ctrl.Result{}, err
+	}
+
+	class, err := r.getGatewayClass(ctx, gateway.Spec.GatewayClassName)
+	if err != nil {
+		return kube_ctrl.Result{}, errors.Wrap(err, "unable to retrieve GatewayClass referenced by Gateway")
```

Comment:
```suggestion
		return kube_ctrl.Result{}, errors.Wrapf(err, "unable to retrieve GatewayClass for %s", req.NamespacedName)
```

As a rule of thumb, I think it's worth including full object names in errors, since it makes it way easier to debug from logs.

Reply:
I think the `req.NamespacedName` is included when the error from `Reconcile` is printed